### ISSUE TITLE
[clojure] warning about unresolved var should not apply to spec code

### ISSUE THREAD

**Issue by allentiak** [syl20bnr/spacemacs#14566]
## Summary

If you write a `defn` *below* the corresponding `s/fdef`, a warning will be given (this should not happen).

If you write a `defn` *above* the corresponding `s/fdef`, no warning will be given (as expected).

## Expected behavior

For specs, the normal "use before definition" principle should not apply, as instrumentation can only be run explicitly.

For instance, the following code works on the terminal:

```clojure
-> % clojure
Clojure 1.10.1
user=> (require '[clojure.spec.alpha :as s]
                '[clojure.spec.test.alpha :as st])
nil

;; We define first the spec for the function...
user=> (s/fdef foo 
          :args (s/cat :a int?))
user/foo

;; Then, the function to be spec'ed
user=> (defn foo [a]         
          (* 2 a))
#'user/foo

;; The only requirement to successfully instrument the function is that both elements have been already defined. The order in which this was done simply does not matter.
user=> (st/instrument `foo)
[user/foo]
user=> (foo "1")
Execution error - invalid arguments to user/foo at (REPL:1).
"1" - failed: int? at: [:a]
user=> (foo 1)
2
```

## Actual behavior

I get a "use-before-def" warning:
!["use-before-def" warning](https://user-images.githubusercontent.com/1922297/110212445-30840380-7e7a-11eb-9277-98eaadde6af7.png)

## Steps to reproduce the problem

Copy the following code into a clojure-mode Emacs buffer, jack-in, and evaluate it.

```clojure
(require '[clojure.spec.alpha :as s]
             '[clojure.spec.test.alpha :as st])

(s/fdef foo 
     :args (s/cat :a int?))

(defn foo [a]         
     (* 2 a))

(st/instrument `foo)

(foo "1")

(foo 1)
```
If you leave the code as-is, with the `defn` *below* the corresponding `s/fdef`, a warning will be given (this should not happen).

![Screenshot -- warning given](https://user-images.githubusercontent.com/1922297/110214488-88733800-7e83-11eb-8e5d-db371c8d902c.png)

However, if you move the `defn` *above* the corresponding `s/fdef`, no warning will be given (as expected).

![Screenshot -- no warning given](https://user-images.githubusercontent.com/1922297/110214486-84dfb100-7e83-11eb-88d7-ddf19b6e5e6b.png)

### CIDER version information
```
;; CIDER 1.1.0snapshot (package: 20210302.849), nREPL 0.8.3
;; Clojure 1.10.1, Java 15.0.2
```

### Emacs version

Spacemacs "develop", running on:

GNU Emacs 27.1 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.23, cairo version 1.16.0) of 2020-11-07, modified by Debian

### Operating system

Debian "testing".

**Comment by allentiak** [comment]
I have confirmed it is *not* a CIDER problem: https://github.com/clojure-emacs/cider/issues/2992

**Comment by allentiak** [comment]
It seems this is a consequence of some interaction across packages. With a minimalistic config, it does not happen...

**Comment by lebensterben** [comment]
as for the "minimalistic config", have you enabled syntax checking layer?

**Comment by allentiak** [comment]
Here are direct links to relevant package lists from the original bug opened for CIDER:
- list of packages unique to my "minimalistic" installation: https://github.com/clojure-emacs/cider/issues/2992#issuecomment-793019224
- the non-shared "long" package list:
https://github.com/clojure-emacs/cider/issues/2992#issuecomment-793021508

**Comment by lebensterben** [comment]
What I mean is if you haven't enabled syntax checking, you should try again with it enabled.

**Comment by allentiak** [comment]
Thanks for your message.

I forgot to specify that, whereas the layer list was different, I kept the
same config for the activated layers in both scenarios. As I executed
exactly the same workflow in both of them, if one scenario had syntax
checking enabled then the other one should have it enabled as well.

That being said we all can make mistakes. So I will let you know if I find
out to be mistaken about this.
